// swagger-typescript-codegen/lib/options/options.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const options_1 = require("./options");
const Mustache = require("mustache");
const defaultOptions = {
    isES6: false,
    moduleName: "",
    includeDeprecated: false,
    imports: [],
    className: "",
    template: {},
    mustache: Mustache,
    beautify: true,
    beautifyOptions: {}
};
describe("makeOptions", () => {
    it("returns the default options when no options are passed", () => {
        const partialOptions = {
            swagger: {}
        };
        expect(options_1.makeOptions(partialOptions)).toEqual(Object.assign({}, defaultOptions, { swagger: {} }));
    });
    it("merges in the options that are passed with higher priority", () => {
        const partialOptions = {
            swagger: {},
            className: "GeneratedDataLayer"
        };
        const options = options_1.makeOptions(partialOptions);
        expect(options.className).toBe("GeneratedDataLayer");
    });
});
describe("validateOptions", () => {
    it("with valid options", () => {
        const partialOptions = {
            swagger: {},
            template: {
                class: "class-template",
                method: "method-template"
            }
        };
        const options = options_1.makeOptions(partialOptions);
        expect(() => options_1.validateOptions(options));
    });
    it("throws when class template is not provided", () => {
        const partialOptions = {
            swagger: {},
            template: {
                class: "class-template",
                method: undefined
            }
        };
        const options = options_1.makeOptions(partialOptions);
        expect(() => options_1.validateOptions(options)).toThrow('Unprovided custom template. Please use the following template: template: { class: "...", method: "...", request: "..." }');
    });
    it("throws when method template is not provided", () => {
        const partialOptions = {
            swagger: {},
            template: {
                class: undefined,
                method: "method-template"
            }
        };
        const options = options_1.makeOptions(partialOptions);
        expect(() => options_1.validateOptions(options)).toThrow('Unprovided custom template. Please use the following template: template: { class: "...", method: "...", request: "..." }');
    });
});
