// swagger-typescript-codegen/lib/options/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Mustache = require("mustache");
const lodash_1 = require("lodash");
const DEFAULT_OPTIONS = {
    isES6: false,
    moduleName: "",
    includeDeprecated: false,
    imports: [],
    className: "",
    template: {},
    mustache: Mustache,
    beautify: true,
    beautifyOptions: {},
    convertQueryParamsToFormDataInPOST: true
};
/**
 * Merge passed options with the default options.
 */
function makeOptions(options) {
    return Object.assign({}, DEFAULT_OPTIONS, options);
}
exports.makeOptions = makeOptions;
/**
 * Validate that the options have required variables for custom generation.
 */
function validateOptions(options) {
    // TODO: Why do we not check for the existence of the type template?
    if (!options.template ||
        !lodash_1.isObject(options.template) ||
        !lodash_1.isString(options.template.class) ||
        !lodash_1.isString(options.template.method)) {
        throw new Error('Unprovided custom template. Please use the following template: template: { class: "...", method: "...", request: "..." }');
    }
}
exports.validateOptions = validateOptions;
