// swagger-typescript-codegen/lib/type-mappers/object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typespec_1 = require("../typespec");
const lodash_1 = require("lodash");
const typescript_1 = require("../typescript");
const any_1 = require("./any");
// see: https://support.reprezen.com/support/solutions/articles/6000162892-support-for-additionalproperties-in-swagger-2-0-schemas
function extractAdditionalPropertiesType(swaggerType, swagger) {
    if (swaggerType.type && swaggerType.type !== "object") {
        return undefined;
    }
    if (swaggerType.additionalProperties === false) {
        return undefined;
    }
    if (swaggerType.additionalProperties === undefined ||
        swaggerType.additionalProperties === true) {
        // is there an easier way to make an "any" type?
        return any_1.makeAnyTypeSpec({
            type: "object",
            required: [],
            minItems: 0,
            title: "any",
            properties: {}
        });
    }
    return typescript_1.convertType(swaggerType.additionalProperties, swagger);
}
exports.extractAdditionalPropertiesType = extractAdditionalPropertiesType;
function makeObjectTypeSpec(swaggerType, swagger) {
    // TODO: We threat everything that reaches this point as an object but not the required properties? (Removing the check for object makes the tests fail)
    // NOTE: object might not be set, because everything without a type is treated as object.
    const requiredPropertyNames = (!swaggerType.type || swaggerType.type === "object") &&
        lodash_1.isArray(swaggerType.required)
        ? swaggerType.required
        : [];
    // Some special handling is needed to support overlapping properties. The list of properties must be reversed to get the
    // overriding properties first. Only then can we filter out any duplicates. To get the original order back, the array
    // is reversed once more
    const allProperties = lodash_1.concat(getAllOfProperties(swaggerType, swagger), getObjectProperties(swaggerType, swagger, requiredPropertyNames));
    const uniqueProperties = lodash_1.uniqBy(lodash_1.reverse(allProperties), "name");
    const properties = lodash_1.reverse(uniqueProperties);
    const addPropsType = extractAdditionalPropertiesType(swaggerType, swagger);
    return Object.assign({}, typespec_1.makeTypeSpecFromSwaggerType(swaggerType), { tsType: "object", isObject: true, isAtomic: false, properties,
        requiredPropertyNames, hasAdditionalProperties: addPropsType !== undefined, additionalPropertiesType: addPropsType });
}
exports.makeObjectTypeSpec = makeObjectTypeSpec;
function getObjectProperties(swaggerType, swagger, requiredPropertyNames) {
    return lodash_1.map(swaggerType.properties, (propertyType, propertyName) => (Object.assign({}, typescript_1.convertType(propertyType, swagger), { name: propertyName, isRequired: lodash_1.includes(requiredPropertyNames, propertyName) })));
}
function getAllOfProperties(swaggerType, swagger) {
    if (!swaggerType.allOf) {
        return [];
    }
    return lodash_1.flatten(lodash_1.map(swaggerType.allOf, ref => {
        if (!ref.$ref) {
            const property = typescript_1.convertType(ref, swagger);
            return lodash_1.filter(property.properties);
        }
        const refSegments = ref.$ref.split("/");
        const name = refSegments[refSegments.length - 1];
        return lodash_1.flatten(lodash_1.filter(lodash_1.map(lodash_1.filter(swagger.definitions, (__, definitionName) => definitionName === name), definition => {
            const property = typescript_1.convertType(definition, swagger);
            return property.properties;
        }), lodash_1.isArray));
    }));
}
