// swagger-typescript-codegen/lib/codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const options_1 = require("./options/options");
const transformToCodeWithMustache_1 = require("./transform/transformToCodeWithMustache");
const swagger2_1 = require("./generators/swagger2");
const enhance_1 = require("./enhance");
exports.CodeGen = {
    transformToViewData: swagger2_1.Swagger2Gen.getViewData,
    transformToCodeWithMustache: transformToCodeWithMustache_1.transformToCodeWithMustache,
    getTypescriptCode: function (opts) {
        const options = options_1.makeOptions(opts);
        return enhance_1.enhanceCode(swagger2_1.Swagger2Gen.getCode(options), options);
    },
    getCustomCode: function (opts) {
        options_1.validateOptions(opts);
        const options = options_1.makeOptions(opts);
        return enhance_1.enhanceCode(swagger2_1.Swagger2Gen.getCode(options), options);
    },
    getDataAndOptionsForGeneration: function (opts) {
        const options = options_1.makeOptions(opts);
        const data = swagger2_1.Swagger2Gen.getViewData(options);
        return { data, options };
    }
};
