// swagger-typescript-codegen/lib/typescript.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = require("./typescript");
const testHelpers_1 = require("./test-helpers/testHelpers");
describe("convertType", () => {
    let swagger;
    let swaggerType;
    let emptyTypeSpecWithDefaults;
    beforeEach(() => {
        swagger = testHelpers_1.makeFakeSwagger();
        emptyTypeSpecWithDefaults = testHelpers_1.makeEmptyTypeSpec();
    });
    describe("reference", () => {
        it("returns a reference object", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                $ref: "https://microsoft.com/api/users",
                type: "reference"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { target: "users", tsType: "ref", isRef: true }));
        });
    });
    describe("enum", () => {
        it("correctly converts an enum type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                enum: ["Marius", "Mark", "Mathieu"],
                type: "enum"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: '"Marius" | "Mark" | "Mathieu"', isAtomic: true, isEnum: true, enum: ["Marius", "Mark", "Mathieu"] }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "namesStartingWithM",
                required: true,
                enum: ["Marius", "Mark", "Mathieu"],
                type: "enum"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "namesStartingWithM", isRequired: true, isNullable: false, tsType: '"Marius" | "Mark" | "Mathieu"', isAtomic: true, isEnum: true, enum: ["Marius", "Mark", "Mathieu"] }));
        });
    });
    describe("string", () => {
        it("correctly converts an string type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({ type: "string" });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "string", isAtomic: true }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a string property",
                required: false,
                type: "string"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a string property", isRequired: false, isNullable: true, tsType: "string", isAtomic: true }));
        });
    });
    describe("number", () => {
        it("correctly converts an number type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({ type: "number" });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "number", isAtomic: true }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a number property",
                required: false,
                type: "number"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a number property", isRequired: false, isNullable: true, tsType: "number", isAtomic: true }));
        });
    });
    describe("boolean", () => {
        it("correctly converts an boolean type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({ type: "boolean" });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "boolean", isAtomic: true }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a boolean property",
                required: false,
                type: "boolean"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a boolean property", isRequired: false, isNullable: true, tsType: "boolean", isAtomic: true }));
        });
    });
    describe("array", () => {
        it("correctly converts an array type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "array",
                items: testHelpers_1.makeSwaggerType({ type: "number" })
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "Array<number>", isAtomic: false, isArray: true, collectionFormat: "multi", elementType: Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "number", isAtomic: true }) }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a array property",
                required: false,
                type: "array",
                collectionFormat: "csv",
                items: testHelpers_1.makeSwaggerType({
                    type: "object",
                    required: false,
                    additionalProperties: false
                })
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a array property", isRequired: false, isNullable: true, isArray: true, collectionFormat: "csv", tsType: "Array<object>", isAtomic: false, elementType: Object.assign({}, emptyTypeSpecWithDefaults, { isRequired: false, isNullable: true, isObject: true, properties: [], tsType: "object", requiredPropertyNames: [] }) }));
        });
    });
    describe("dictionary", () => {
        it("correctly converts an dictionary type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                additionalProperties: testHelpers_1.makeSwaggerType({ type: "number" })
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: [], hasAdditionalProperties: true, additionalPropertiesType: Object.assign({}, emptyTypeSpecWithDefaults, { isAtomic: true, tsType: "number" }) }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a dictionary property",
                required: false,
                type: "object",
                additionalProperties: testHelpers_1.makeSwaggerType({ type: "number" })
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a dictionary property", tsType: "object", isAtomic: false, isObject: true, isNullable: true, isRequired: false, properties: [], requiredPropertyNames: [], hasAdditionalProperties: true, additionalPropertiesType: Object.assign({}, emptyTypeSpecWithDefaults, { isAtomic: true, tsType: "number" }) }));
        });
    });
    describe("any", () => {
        it("correctly converts an any type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                minItems: 10,
                title: "Een mooie titel"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "any", isAtomic: true }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a any property",
                required: false,
                type: "object",
                minItems: 10,
                title: "Een mooie titel"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a any property", isRequired: false, isNullable: true, tsType: "any", isAtomic: true }));
        });
    });
    describe("object", () => {
        it("correctly converts an object type", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: [] }));
        });
        it("correctly passes through typespec properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                description: "The description of a any property",
                required: false,
                type: "object"
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { description: "The description of a any property", isRequired: false, isNullable: true, tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: [], hasAdditionalProperties: true, additionalPropertiesType: Object.assign({}, emptyTypeSpecWithDefaults, { isAtomic: true, tsType: "any" }) }));
        });
        it("correctly converts an object type with properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                properties: {
                    age: testHelpers_1.makeSwaggerType({ type: "number" })
                },
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [
                    Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "number", isAtomic: true, name: "age", isRequired: false })
                ], requiredPropertyNames: [] }));
        });
        it("correctly converts an object type with additionalProperties: false", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: [] }));
        });
        it("correctly converts an object type with additionalProperties: true", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                additionalProperties: true
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: [], hasAdditionalProperties: true, additionalPropertiesType: Object.assign({}, emptyTypeSpecWithDefaults, { isAtomic: true, tsType: "any" }) }));
        });
        it("correctly converts an object type with a missing additionalProperties field", () => {
            // this needs to be treated exactly the same as with "additionalProperties = true"
            // see: https://support.reprezen.com/support/solutions/articles/6000162892-support-for-additionalproperties-in-swagger-2-0-schemas
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                additionalProperties: undefined
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: [], hasAdditionalProperties: true, additionalPropertiesType: Object.assign({}, emptyTypeSpecWithDefaults, { isAtomic: true, tsType: "any" }) }));
        });
        it("handles required properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                properties: {
                    age: testHelpers_1.makeSwaggerType({ type: "number" })
                },
                required: ["age"],
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [
                    Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "number", isAtomic: true, name: "age", isRequired: true })
                ], requiredPropertyNames: ["age"] }));
        });
        // TODO: This behaviour seems kind of weird? There is an object in an object but we seem flatten them in the resulting typedef?
        // TODO: It also seems weird that the requiredPropertiesNames does not take into account the flattening
        it("handles allOf properties of non $ref properties", () => {
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                allOf: [
                    testHelpers_1.makeSwaggerType({
                        type: "object",
                        properties: {
                            age: testHelpers_1.makeSwaggerType({ type: "number" })
                        },
                        additionalProperties: false
                    })
                ],
                required: ["age"],
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [
                    Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "number", isAtomic: true, name: "age", isRequired: false })
                ], requiredPropertyNames: ["age"] }));
        });
        it("handles allOf properties of $ref properties", () => {
            swagger = Object.assign({}, swagger, { definitions: {
                    person: testHelpers_1.makeSwaggerType({
                        type: "object",
                        properties: {
                            age: testHelpers_1.makeSwaggerType({ type: "number" })
                        },
                        additionalProperties: false
                    })
                } });
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                allOf: [
                    testHelpers_1.makeSwaggerType({
                        $ref: "api/person",
                        type: "reference"
                    })
                ],
                required: ["age"],
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [
                    Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "number", isAtomic: true, name: "age", isRequired: false })
                ], requiredPropertyNames: ["age"] }));
        });
        it("does not handle allOf properties of $ref properties that do not have a definition", () => {
            swagger = Object.assign({}, swagger, { definitions: {
                    person: testHelpers_1.makeSwaggerType({
                        type: "object",
                        properties: {
                            age: testHelpers_1.makeSwaggerType({ type: "number" })
                        },
                        additionalProperties: false
                    })
                } });
            swaggerType = testHelpers_1.makeSwaggerType({
                type: "object",
                allOf: [
                    testHelpers_1.makeSwaggerType({
                        $ref: "api/location",
                        type: "reference"
                    })
                ],
                required: ["age"],
                additionalProperties: false
            });
            expect(typescript_1.convertType(swaggerType, swagger)).toEqual(Object.assign({}, emptyTypeSpecWithDefaults, { tsType: "object", isAtomic: false, isObject: true, properties: [], requiredPropertyNames: ["age"] }));
        });
    });
});
